book/src/dev/rfcs/XXXX-asynchronous-script-verification.md

```diff
@@ -0,0 +1,231 @@
+
+1. behaviors related to the state's external API (a `Buffer`ed `tower::Service`);
+2. behaviors related to the state's internal implementation (using `sled`).
+
+Making this distinction helps us to ensure we don't accidentally leak
+"internal" behaviors into "external" behaviors, which would violate
+encapsulation and make it more difficult to replace `sled`.
+
+In the first category, our state is presented to the rest of the application
+as a `Buffer`ed `tower::Service`. The `Buffer` wrapper allows shared access
+to a service using an actor model, moving the service to be shared into a
+worker task and passing messages to it over an multi-producer single-consumer
+(mpsc) channel. The worker task recieves messages and makes `Service::call`s.
+The `Service::call` method returns a `Future`, and the service is allowed to
+decide how much work it wants to do synchronously (in `call`) and how much
+work it wants to do asynchronously (in the `Future` it returns).
+
+This means that our external API ensures that the state service sees a
+linearized sequence of state requests, although the exact ordering is
+unpredictable when there are multiple senders making requests.
+
+In the second category, the Sled API presents itself synchronously, but
+database and tree handles are clonable and can be moved between threads. All
+that's required to process some request asynchronously is to clone the
+appropriate handle, move it into an async block, and make the call as part of
+the future. (We might want to use Tokio's blocking API for this, but that's a
+side detail).
+
+Because the state service has exclusive access to the sled database, and the
+state service sees a linearized sequence of state requests, we have an easy
+way to opt in to asynchronous database access. We can perform sled operations
+synchronously in the `Service::call`, waiting for them to complete, and be
+sure that all future requests will see the resulting sled state. Or, we can
+perform sled operations asynchronously in the future returned by
+`Service::call`.
+
+If we perform all *writes* synchronously and allow reads to be either
+synchronous or asynchronous, we ensure that writes cannot race each other.
+Asynchronous reads are guaranteed to read at least the state present at the
+time the request was processed, or a later state.
+
+Now, returning to the UTXO lookup problem, we can map out the possible states
+with this restriction in mind. This description assumes that UTXO storage is
+split into disjoint sets, one in-memory (e.g., blocks after the reorg limit)
+and the other in sled (e.g., blocks after the reorg limit). The details of
+this storage are not important for this design, only that the two sets are
+disjoint.
+
+When the state service processes a `Request::AwaitUtxo(OutPoint)` referencing
+some UTXO `u`, there are three disjoint possibilities:
+
+1. `u` is already contained in an in-memory block storage;
+2. `u` is already contained in the sled UTXO set;
```

Comment:
Do we need to concern ourselves with how the state struct is managing the commited state? I think from the perspective of the pending utxos we might be able to simplify this to "in the state" and "not in the state" without unnecessarily forcing the lookups to serialize in a way that hurts the common case with old utxos.

Reply:
This is supposed to be relatively agnostic to how the actual state structure is organized, the point is just to divide between data that's in sled and data that isn't; we have to check both.

Reply:
> This is supposed to be relatively agnostic to how the actual state structure is organized, the point is just to divide between data that's in sled and data that isn't; we have to check both.

Agreed. But the fact that we don't care how the state structure is organized implies that it is irrelevant here and shouldn't be mentioned as part of this RFC.

Reply:
Hmm, but it does seem important to name the difference, because we shouldn't be checking both of them the same way: the in-memory data is small and can be accessed synchronously, while the UTXO set should be queried async since it might require reading from disk.  That's the distinction I was trying to highlight.